torch/distributed/fsdp/flat_param.py

```diff
@@ -846,6 +846,36 @@ def _free_low_precision_sharded_param(self):
         self._check_low_precision_shard()
         _free_storage(self.flat_param._mp_shard)  # type: ignore[attr-defined]
 
+    @torch.no_grad()
+    def unshard_grad(self):
+        if not self.uses_sharded_strategy:
+            return
+        flat_param = self.flat_param
+        self._check_unsharded(flat_param)
+        padded_unsharded_grad = torch.empty(
+            flat_param._padded_unsharded_size,  # type: ignore[attr-defined]
+            device=self.device,
+        )
+        if flat_param.grad is None:
+            flat_param._saved_grad_shard = None  # type: ignore[attr-defined]
+            sharded_grad = torch.zeros_like(flat_param)  # type: ignore[attr-defined]
+        else:
+            self._check_sharded(flat_param.grad)
+            flat_param._saved_grad_shard = flat_param.grad  # type: ignore[attr-defined]
+            sharded_grad = flat_param._saved_grad_shard  # type: ignore[attr-defined]
+        dist._all_gather_base(padded_unsharded_grad, sharded_grad, self.process_group)
```

Comment:
We discussed briefly offline, but there are two options:
1. (As in the PR currently) We only use a single all-gather collective per `FlatParameter`. In that case, if all ranks' sharded gradient is `None`, then the unsharded gradient is incorrectly `torch.zeros(unsharded_size)`. If only some ranks' sharded gradients are `None`, then the unsharded gradient zeros those corresponding elements.
2. We use a preceding all-reduce collective per `FlatParameter` to indicate if each rank's sharded gradient is `None` or not. This solves the problem from 1.

Since `summon_full_params(with_grads=True)` is meant for debugging, I can see the argument for pursuing 2. I can change this in a follow-up PR and add/adjust unit tests accordingly.